### ISSUE TITLE
PSP is not working as expected in my test scenario

### ISSUE THREAD

**Issue by nizamettinOzpolat** [MicrosoftDocs/sql-docs#7849]
Hello,

I wanted to test Parameter Sensitive Plan. I created a table with 80.000 rows of data. 40.000 rows has the same id, 60,000, the rest is incremental, from 1 to 40.000. Here is the repro script.

Create database TestDB_FOR_PSP
go
USE [master]
GO
ALTER DATABASE [TestDB_FOR_PSP] SET QUERY_STORE = ON
GO
ALTER DATABASE [TestDB_FOR_PSP] SET QUERY_STORE (OPERATION_MODE = READ_WRITE)
GO
ALTER DATABASE [TestDB_FOR_PSP] SET QUERY_STORE (QUERY_CAPTURE_MODE = auto);
go
use TestDB_FOR_PSP
go

Create table DBO.T1(id int identity(1,1), c1 varchar(50))
go
insert into dbo.T1 (c1) values('rwrw ererwe ewr')
go 40000
Create table DBO.T2(id int, c1 varchar(50))
go
insert into dbo.T2
select * from dbo.T1
go

insert into dbo.T2
select top 40000 60000 as id, c1 from dbo.T1

create index idx on dbo.t2 (id)

--select * from dbo.T2
go
--include actal execution plani (ctrt+M)
--Index seek operator is expected in execution plan, 
exec sp_executesql N'select * from dbo.t2 where  id=@id',N'@id int', @id = 300
go
--Index scan operator is expected in the plan, but index seek is used
exec sp_executesql N'select * from dbo.t2 where  id=@id',N'@id int', @id = 60000
go
![image](https://user-images.githubusercontent.com/90388408/180445605-b16aa706-1ce5-48f7-9f73-457e13987b8d.png)

-- plan cach is cleared
dbcc freeproccache

--Index scan operator is expected in the plan.
exec sp_executesql N'select * from dbo.t2 where  id=@id',N'@id int', @id = 60000
go
--Index seek operator is expected in execution plan,  but index scan is used
exec sp_executesql N'select * from dbo.t2 where  id=@id',N'@id int', @id = 300
go
![image](https://user-images.githubusercontent.com/90388408/180445712-267d9b86-f532-4c6d-9d71-0b21ecd68ae1.png)

Best regards
Nizamettin Ozpolat
---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: d912455d-8d09-0b74-0f8e-9de4d8777b26
* Version Independent ID: d912455d-8d09-0b74-0f8e-9de4d8777b26
* Content: [Parameter Sensitivity Plan optimization - SQL Server](https://docs.microsoft.com/en-us/sql/relational-databases/performance/parameter-sensitivity-plan-optimization?view=sql-server-ver16)
* Content Source: [docs/relational-databases/performance/parameter-sensitivity-plan-optimization.md](https://github.com/MicrosoftDocs/sql-docs/blob/live/docs/relational-databases/performance/parameter-sensitivity-plan-optimization.md)
* Product: **sql**
* Technology: **performance**
* GitHub Login: @thesqlsith
* Microsoft Alias: **derekw**

**Comment by thesqlsith** [comment]
Thank you for reporting this.  The main reason why PSP did not kick in for this particular example is that there was not enough data for PSP to consider how skewed the data was.  In the current CTP for SQL Server 2022, the feature will check for the high range boundary as can be seen in the ShowPlan_XML for an executed query (i.e. <ParameterSensitivePredicate LowBoundary="100" HighBoundary="100000">) before it will dig a little further into the data skewness.  This is not currently documented this way as the HighBoundary may change before we release SQL Server 2022.  

As a result, the second query just reused the index seek plan since that plans' parameter was "sniffed" from the first select query.

I have made a quick modification to your original example and added a few comments that I hope will be useful.

USE [master]
GO

Create database TestDB_FOR_PSP
go

ALTER DATABASE [TestDB_FOR_PSP] SET QUERY_STORE (OPERATION_MODE = READ_WRITE, DATA_FLUSH_INTERVAL_SECONDS = 60, INTERVAL_LENGTH_MINUTES = 1, QUERY_CAPTURE_MODE = AUTO)
GO

ALTER DATABASE [TestDB_FOR_PSP] SET RECOVERY SIMPLE
GO

use TestDB_FOR_PSP
go

ALTER DATABASE TestDB_FOR_PSP SET QUERY_STORE CLEAR ALL;
ALTER DATABASE SCOPED CONFIGURATION CLEAR PROCEDURE_CACHE;
GO

DROP TABLE IF EXISTS DBO.T1;
GO
DROP TABLE IF EXISTS DBO.T2;
GO

Create table DBO.T1(id int identity(1,1), c1 varchar(4000))
go

SET NOCOUNT ON
GO

INSERT INTO DBO.T1 (c1)
--SELECT TOP 40000 [text]

--create a Cartesian product out of the messages table
--to speed up the insert process
select top 40000 a1.text as a1_text
from sys.messages a1, sys.messages a2;
GO
/*Uncomment the next three lines to add more data and comment out the three lines above
select top 200000 a1.text as a1_text
from sys.messages a1, sys.messages a2
GO
*/

--insert into dbo.T1 (c1) values('rwrw ererwe ewr')
--go 40000

Create table DBO.T2(id int, c1 varchar(4000))
go

insert into dbo.T2
select * from dbo.T1
go

insert into dbo.T2
select top 40000 60000 as id, c1 from dbo.T1

/* Uncomment the next two lines and comment out the above select top statement in order to work with a slightly larger dataset
select top 100000 60000 as id, c1 from dbo.T1;
create index idx on dbo.t2 (id)
*/
--select * from dbo.T2
go

-- See stats, 3 RANGE_HI_KEY values
--DBCC SHOW_STATISTICS ('dbo.t2', 'idx');

SELECT sh.*, sp.[rows], sp.rows_sampled, sp.last_updated
FROM sys.stats AS s
CROSS APPLY sys.dm_db_stats_histogram(s.object_id, s.stats_id) AS sh
CROSS APPLY sys.dm_db_stats_properties (s.object_id, s.stats_id) AS sp
WHERE s.name = 'idx' AND s.object_id = OBJECT_ID('dbo.t2')
GO


--Setup XE capture
IF EXISTS (SELECT * FROM sys.server_event_sessions WHERE name = 'PSP')
DROP EVENT SESSION [TEST_PSP_XE] ON SERVER;
GO
CREATE EVENT SESSION [TEST_PSP_XE] ON SERVER 
ADD EVENT sqlserver.parameter_sensitive_plan_optimization(
    ACTION(sqlserver.query_hash_signed,sqlserver.query_plan_hash_signed,sqlserver.sql_text)),
ADD EVENT sqlserver.parameter_sensitive_plan_testing(
    ACTION(sqlserver.query_hash_signed,sqlserver.query_plan_hash_signed,sqlserver.sql_text)),
ADD EVENT sqlserver.parameter_sensitive_plan_optimization_skipped_reason(
    ACTION(sqlserver.query_hash_signed,sqlserver.query_plan_hash_signed,sqlserver.sql_text))
WITH (MAX_MEMORY=4096 KB,EVENT_RETENTION_MODE=NO_EVENT_LOSS,MAX_DISPATCH_LATENCY=1 SECONDS,MAX_EVENT_SIZE=0 KB,MEMORY_PARTITION_MODE=NONE,TRACK_CAUSALITY=OFF,STARTUP_STATE=OFF)
GO

-- Start XE
ALTER EVENT SESSION [TEST_PSP_XE] ON SERVER
STATE = START;
GO

-- Take a look at the live view XE session to see if PSP is kicking in or not.  There
-- will be a reason provided if we decided to skip engaging PSP

--include actal execution plani (ctrt+M)
--Index seek operator is expected in execution plan
SET STATISTICS XML ON;
exec sp_executesql N'select * from dbo.t2 where id=@id',N'@id int', @id = 300
SET STATISTICS XML OFF;
go
--Index scan operator is expected in the plan, but index seek is used
SET STATISTICS XML ON;
exec sp_executesql N'select * from dbo.t2 where id=@id',N'@id int', @id = 60000
SET STATISTICS XML OFF;
GO

--Look at the plan cache, the addtional hints that PSP adds to a PSP eligible query
--will show up in the text and ShowPlan_XML
SELECT cp.usecounts,  
    SUBSTRING(qt.text,qs.statement_start_offset/2 +1,   
                 (CASE WHEN qs.statement_end_offset = -1   
                       THEN LEN(CONVERT(nvarchar(max), qt.text)) * 2   
                       ELSE qs.statement_end_offset end -  
                            qs.statement_start_offset  
                 )/2  
             ) AS query_text,
    qs.query_hash,
	qs.query_plan_hash,
	cp.plan_handle,
	qs.sql_handle,
	qp.query_plan,
	qt.text
FROM sys.dm_exec_cached_plans AS cp
INNER JOIN sys.dm_exec_query_stats AS qs
	ON cp.plan_handle = qs.plan_handle
CROSS APPLY sys.dm_exec_sql_text(qs.sql_handle) AS qt
CROSS APPLY sys.dm_exec_query_plan(qs.plan_handle) AS qp
WHERE qt.text like '%select * from dbo.t2 where%'   
ORDER BY qs.execution_count DESC; 
GO

--Because this example is using sp_executesql versus a stored procedure
--a second execution of the query that produced a query variant may be
--needed to see it in Query Store.  The variant can be seen in the plan
--cache and the showplan_xml without a second execution however.  The XE
--session should also show where PSP kicked in without a second execution
exec sp_executesql N'select * from dbo.t2 where id=@id',N'@id int', @id = 60000
go

--Look at the variant info in QDS
SELECT  Pl.query_id , qvr.query_variant_query_id,Txt.query_sql_text, Pl.plan_id,Pl.query_id ,qvr.parent_query_id,qvr.dispatcher_plan_id,
convert(xml,Pl.query_plan)as ShowPlanXML, Rs.first_execution_time, Rs.last_execution_time, Rs.count_executions AS NumberOfExecutions, 
RS.avg_rowcount, Rs.max_rowcount, Qry.count_compiles AS NumberOfCompiles, Qry.initial_compile_start_time, Qry.last_compile_start_time, 
Qry.last_execution_time,Qry.query_hash
FROM sys.query_store_runtime_stats AS Rs
right outer JOIN sys.query_store_plan AS Pl on Rs.plan_id = Pl.plan_id
left outer join sys.query_store_query_variant qvr on Pl.query_id = qvr.query_variant_query_id
JOIN sys.query_store_query AS Qry  
    ON Pl.query_id = Qry.query_id  
JOIN sys.query_store_query_text AS Txt  
    ON Qry.query_text_id = Txt.query_text_id  
where (Pl.query_id  in (select parent_query_id from sys.query_store_query_variant ) or qvr.parent_query_id is not null)
order by query_hash, Pl.query_id;

Using the original code the PSP Extended Event (parameter_sensitive_plan_optimization_skipped_reason) will show that the data Skewness Threshold was not met.
![image](https://user-images.githubusercontent.com/34077013/180492484-73ad0c13-b33b-49ad-85ab-91ca1b680333.png)

If you re-run the query and uncomment the lines where more data is added to the dataset then PSP will kick in and you will see the index scan for the second select statement.
